src/domain/usecases/SendNotificationDataValuesUseCase.ts

```diff
@@ -0,0 +1,347 @@
+            })
+            .flatten()
+            .value();
+
+        const dataSets = await this.dataSetsRepository.getBy(dataSetKeys);
+        const orgUnits = await this.orgUnitRepository.get(orgUnitIds);
+        const dataSetStore = await this.dataSetStoreRepository.get({
+            path: options.jsonExecutionsPath,
+            namespace: options.dataStoreNameSpace,
+        });
+
+        const dataSetExecutions = this.getDataSetsExecutions(
+            dataSetKeys,
+            settings,
+            dataSets,
+            orgUnits,
+            dataSetStore
+        );
+
+        if (_.isEmpty(dataSetStore)) {
+            await this.saveExecutions(dataSetExecutions, options);
+        }
+
+        await promiseMap(_(dataSetExecutions).keys().value(), async key => {
+            const execution = dataSetExecutions[key];
+            if (!execution) return false;
+            const [dataSetId, orgUnitId, period] = key.split("-");
+            if (!orgUnitId || !dataSetId || !period) {
+                throw Error("Cannot found org unit, dataset or period in execution");
+            }
+
+            const dataValues = await this.dataValuesRepository.get({
+                includeDeleted: false,
+                dataSetIds: [dataSetId],
+                orgUnitIds: [orgUnitId],
+                periods: [period],
+                lastUpdated: execution.lastUpdated,
+            });
+            log.debug(`Getting ${dataValues.length} datavalues for execution: ${key}`);
+
+            const userGroup = this.getDataSetConfig(dataSets, dataSetId, settings)?.userGroup;
+            if (this.diffSinceLastUpdate(dataValues) >= options.sendEmailAfterMinutes && userGroup) {
+                const dataValuesWithOldValue = await this.getOldDataValues(
+                    dataSetId,
+                    orgUnitId,
+                    period,
+                    dataValues
+                );
+
+                const users = await this.getUsersToNotify(dataValues, userGroup);
+                log.debug(`Sending email to ${users.length} users in group ${userGroup}`);
+
+                const { body, subject } = await this.generateEmailTemplate(
+                    dataSets,
+                    dataSetId,
+                    orgUnits,
+                    orgUnitId,
+                    dataValuesWithOldValue,
+                    options,
+                    period
+                );
+
+                await this.sendEmailToUsersInGroup(users, subject, body);
+
+                await this.saveExecutionsWithEmail(execution, dataSetExecutions, key, options);
+                log.debug(`Executions saved: ${key}`);
+            }
+        });
+    }
+
+    private async generateEmailTemplate(
+        dataSets: DataSet[],
+        dataSetId: string,
+        orgUnits: OrgUnit[],
+        orgUnitId: string,
+        dataValuesWithOldValue: DataValue[],
+        options: OptionsUseCase,
+        period: string | undefined
+    ) {
```

Comment:
[best practices] Be explicit about the return type of a function, so a reader knows what's it's returning without hovering the method/inspecting the code. Also, if a bug is introduced, the error will show here and not on the caller code. [Same for all methods in the class]